refactor: migrate to motor package

chooyan-eng opened the pull request:
close #11

migration from [springster](https://pub.dev/packages/springster) to [motor](https://pub.dev/packages/motor) 

## Detail

- replace `springster` into `motor`
- replace `SpringSimulationController2D` to `PolyMotionController`
- `SpringDescription` is converted into `Motion`
- default `SpringDescription` is replaced from `Spring.defaultIOS` into `SpringDescription.withDurationAndBounce()`

chooyan-eng committed:
refactor: migrate to motor package

chooyan-eng commented:
@timcreatedit Could you check this PR whether I'm using `motor` correctly if possible? I've already confirmed every example page works perfectly.

timcreatedit commented on example/lib/spring_page.dart:
```diff
@@ -22,9 +22,12 @@ class _SpringPageState extends State<SpringPage> with TickerProviderStateMixin {
+        SpringType.bouncy =>
+          (Motion.bouncySpring() as SpringMotion).description,
+        SpringType.snappy =>
+          (Motion.snappySpring() as SpringMotion).description,
```
You can probably avoid the type casting like this:

```suggestion
         CupertinoMotion.smooth().description,
        SpringType.bouncy =>
          CupertinoMotion.bouncy().description,
        SpringType.snappy =>
          CupertinoMotion.snappy().description,
```

chooyan-eng replied:
oops, I didn't find this! Thanks!

chooyan-eng replied:
fixed 1bd0a2bd095c6cdfcc48305c282b38a998d308df

timcreatedit commented on lib/src/widget.dart:
```diff
@@ -188,7 +187,7 @@ class AnimatedTo extends StatelessWidget {
         )
       : SpringAnimatedTo(
           globalKey: globalKey,
-          description: description ?? Spring.defaultIOS,
+          description: description ?? SpringDescription.withDurationAndBounce(),
```
I think in the future it would be nice to let users pass any kind of motion, but I also understand if you don't want to expose `motor`s internals here.

chooyan-eng replied:
Yeah, it's one of my highest priority to keep my packages independent from others, to be honest. I'm thinking of introducing `motor` in the README, as in the example app uses it.

timcreatedit replied:
I really get that, maybe we should gonna release motor_animated_to or something that bridges the gap between the two :)

timcreatedit commented on lib/src/spring_widget.dart:
```diff
@@ -174,9 +176,10 @@ class _RenderAnimatedTo extends RenderProxyBox {
       vsync: _vsync,
-      spring: description,
+      initialValue: [0, 0],
     )..addListener(_attemptPaint);
```
Using `PolyMotionController` is technically correct here and it will work. However, the big advantage of motor is that you could also use `MotionController<Offset>` and pass an `MotionConverter.offset` in the constructor. That way you don't have to keep track of array values like this and can work with `Offset` directly :)

chooyan-eng replied:
Thanks, I'll try that 💪

chooyan-eng replied:
@timcreatedit I tried `MotionController<Offset>` at 1bd0a2bd095c6cdfcc48305c282b38a998d308df, and now the source code looks so clean and safe! I now understand why `PolyMotionController` is not exported from the library.

timcreatedit commented on lib/src/spring_widget.dart:
```diff
@@ -189,7 +192,7 @@ class _RenderAnimatedTo extends RenderProxyBox {
 
   set description(SpringDescription value) {
-    _controller.spring = value;
+    _controller.motion = Motion.customSpring(value);
```
Depending on the desired behavior, you might want to do this:

```suggestion
    _controller.motion = SpringMotion(value, snapToEnd: true);
```

This will ensure that the animation settles at the exact target value, but it might introduce a small jump at the very end. 

The version of `springster` that you used before did not have this behavior, so if you keep it the way you have it, everything will stay the same as it was before.

In a future version of motor, I will default this value to true, as having it set to false can introduce some unpredictable behavior sometimes. If you want me to elaborate on my reasoning for that I can do that, but I think for `animated_to` it's not that important.

chooyan-eng replied:
I guess I have to compare the behavior first, but as long as reading your comment, `animated_to` should use `snapToEnd: true` as it may be more important for package users to stop at the exact position they configure as their widget structure.

chooyan-eng replied:
I've compared and didn't find any problem with jumping at the end, so I've fixed using `snapToEnd: true` as you suggested. Thanks.

timcreatedit commented on lib/src/spring_widget.dart:
```diff
@@ -174,9 +176,10 @@ class _RenderAnimatedTo extends RenderProxyBox {
         _horizontalScrollOffset = horizontalScrollOffset {
-    _controller = SpringSimulationController2D.unbounded(
+    _controller = PolyMotionController(
+      motion: Motion.customSpring(description),
```
See my other comment regarding `snapToEnd`

chooyan-eng commented:
@timcreatedit Thank you for all the comments and suggestions in detail! I'll fix them!

chooyan-eng committed:
refactor: improve motor integration based on PR feedback

- Use MotionController<Offset> with MotionConverter.offset instead of PolyMotionController. This doesn't require List<double> for (x, y) anymore.
- Use snapToEnd: true to SpringMotion for precise positioning at target values
- Remove unnecessary type casting in example (use CupertinoMotion directly)
- Simplify velocity type from List<double> to Offset for better type safety

Addresses PR review comments from @timcreatedit

chooyan-eng commented:
@timcreatedit Again, thank you for your review! Now the migration looks way better than my first commit! I'll merge this and publish it after updating README. If it still has some points to be fixed or considered, kindly let me know. 👍
